lib/Sema/TypeCheckAvailability.cpp

```diff
@@ -3496,6 +3496,20 @@ class ExprAvailabilityWalker : public ASTWalker {
     return call;
   }
 
+  /// Walks up to the first enclosing LoadExpr and returns it.
+  const LoadExpr *getEnclosingLoadExpr() const {
+    assert(!ExprStack.empty() && "must be called while visiting an expression");
+    ArrayRef<const Expr *> stack = ExprStack;
+    stack = stack.drop_back();
+
+    for (auto expr : llvm::reverse(stack)) {
+      if (auto loadExpr = dyn_cast<LoadExpr>(expr))
+        return loadExpr;
+    }
```

Comment:
I think you may also want to check that the lvalue is in the base of a parent SubscriptExpr, e.g this should still warn:

```swift
struct S {
    var x: Int

    static var bar: S {
        get { S(x: 0) }
        @available(*, deprecated)
        set { }
    }
    mutating func foo() -> Int {
        x = 1
        return x
    }
}

var arr = [0, 1]
let y = arr[S.bar.foo()]
```

I guess KeyPathApplicationExpr probably ought to be checked for too (although I don't think lvalues can currently occur in their argument since we ban inout subscripts for key paths).

Reply:
That's a good test case, thanks!

Reply:
So as it turns out, all the test cases I can come up with involving keypaths are broken at the moment (we don't seem to ever emit accessor availability diagnostics when forming keypaths). This is worth following up on, and I've now got test cases demonstrating the brokenness, but I'm not gong to try to fix it in this PR because fixing missing diagnostics is a lot riskier than fixing diagnostics that shouldn't be emitted, source compatibility-wise.

Reply:
I handled some of the test cases you brought up in this follow-on PR: https://github.com/apple/swift/pull/72410